mitodl/mit-open
# Normalize learning resource serializers

# What are the relevant tickets?
Closes #36 

# Description (What does it do?)

- Refactored the serializers into some smaller packages
- Updated `LearningResourceSerializer` to be a polymorphic wrapper around the resource_type specific variants. This now allows the API schema and thus the typescript generated code to be type specific in places where we know what the type is.

**Main branch:**
![Screenshot 2023-10-20 at 00-14-59 MIT Open Discussions Course Catalog API](https://github.com/mitodl/mit-open/assets/28598/bce35b84-d5d6-49d5-9bf0-3733c1d893d7)

**This Branch:**
![Screenshot 2023-10-20 at 00-11-29 MIT Open Discussions Course Catalog API](https://github.com/mitodl/mit-open/assets/28598/54e84362-12d7-4325-959a-15773c47bb56)

- Switched `ProgramResourceSerializer.program.courses` to be course learning resources (note the is now typed to `CourseResource`, not a generic LearningResource):

**Main branch:**
![Screenshot 2023-10-20 at 00-16-32 MIT Open Discussions Course Catalog API](https://github.com/mitodl/mit-open/assets/28598/a695fe03-936a-4243-bc63-6d1238302d86)


**This Branch:**
![Screenshot 2023-10-20 at 00-12-05 MIT Open Discussions Course Catalog API](https://github.com/mitodl/mit-open/assets/28598/d0aeacaf-771d-4432-8c87-97c197985962)

- All type-specific APIs now return a specific type using that type's serializer. For example, the programs API:

**Main branch:**
![Screenshot 2023-10-20 at 00-19-45 MIT Open Discussions Course Catalog API](https://github.com/mitodl/mit-open/assets/28598/c2028c13-8341-4855-b613-e7714bf55d2c)

**This Branch:**
![Screenshot 2023-10-20 at 00-20-13 MIT Open Discussions Course Catalog API](https://github.com/mitodl/mit-open/assets/28598/ed31aada-d942-4043-9818-fbdb906bea08)

- Updated `factoryboy` and the factory definitions themselves so that the test data they generate is more internally consistent. Tests were breaking otherwise.

# How can this be tested?

- Check out the schema and make sure it makes sense compared to what's on main/RC.
- Test the API in the swagger UI to verify it's returning the revised data structures.
- Functional test the application.
